tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,258 @@
+    print(f'Processing autogen input file: {autogen_csv}')
+    autogen_dfs.append(pd.read_csv(autogen_csv).fillna(""))
+  if autogen_dfs:
+    df_svs = pd.concat([df_svs] + autogen_dfs)
+    df_svs = df_svs.drop_duplicates(subset=utils.DCID_COL)
+
+  # Get alternatives and add to the dataframe.
+  for alt_fp in sorted(glob.glob(alternatives_filepattern)):
+    df_alts = utils.get_local_alternatives(
+        alt_fp, [utils.DCID_COL, utils.ALTERNATIVES_COL])
+    df_svs = utils.merge_dataframes(df_svs, df_alts)
+
+  return df_svs
+
+
+def fine_tune_model(model: Any, df_svs: pd.DataFrame,
+                    df_sentence_pairs: pd.DataFrame):
+  """Fine tuning involves providing pairs of sentences/text with an
+  approximate similar score to a baseline model. These pairs (and scores)
+  are used for further `training` (with CosineSimilarityLoss as the objecive).
+  The end result is that the baseline model's weights get updated based on
+  the `new` training examples (pairs).
+  Effectively, the training examples (pairs and scores) provide additional
+  context to a baseline model about the kinds of associations between
+  sentences/text that we care about. In this case, we use the StatVar name,
+  description and alternatives (human curated and LLM-generated) to create
+  pairs of sentences/text with high similarity scores. We also provide a
+  way to indicate that some pairs of texts/words/sentences should not be
+  associated with each other by assigning very low scores. This allows us to
+  bias the model's weights to accomodate our particular use case better. 
+  """
+
+  training_examples: List[InputExample] = []
+
+  # Iterate over SV Names, Descriptions and Alternatives to produce
+  # pairs of texts which we want to associated with each other in
+  # terms of similarity.
+  for _, row in df_svs.iterrows():
+    name = row[utils.NAME_COL]
+    descriptions = row[utils.DESCRIPTION_COL].split(";")
+    desc = descriptions[0]
+
+    if not name and not desc:
+      continue
+    if ((not name) and desc):
+      name = desc
+    elif ((not desc) and name):
+      desc = name
+
+    curated = row[utils.CURATED_ALTERNATIVES_COL].split(";")
+    palm_alts = row[utils.ALTERNATIVES_COL].split(";")
+
+    # Pair name and description with very high score.
+    if name != desc:
+      training_examples.append(
+          InputExample(texts=[name, desc], label=VERY_HIGH_MATCH_SCORE))
+
+    # If there are more descriptions, pair them with the name as well.
+    if len(descriptions) > 1:
+      for i in range(1, len(descriptions)):
+        if descriptions[i]:
+          training_examples.append(
+              InputExample(texts=[name, descriptions[i]],
+                           label=HIGH_MATCH_SCORE))
+
+    for c in curated:
+      # All all curated alternatives as pairs with the description.
+      if c:
+        training_examples.append(
+            InputExample(texts=[desc, c], label=HIGH_MATCH_SCORE))
```

Comment:
Any reason why curated uses `desc` and not `name`?

Also, wonder why not do VERY_HIGH_MATCH_SCORE for the curated descriptions, since manually curated.

Reply:
No reason for using `desc` vs `name`. Just going for some variety. Good point about the VERY_HIGH_MATCH_SCORE. Using that.